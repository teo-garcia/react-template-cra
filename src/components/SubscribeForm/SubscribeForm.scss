.subscribe-form {
  display: flex;
  flex-flow: column;
  margin: 32px auto 72px;

  @media (min-width: 1024px) {
    width: 500px;
    flex-flow: row;
    flex-wrap: wrap;
    align-items: center;
    justify-content: space-between;
  }

  &__label {
    font-weight: var(--font-weight-medium);
    font-size: 1.4rem;
    padding-left: 2px;

    @media (min-width: 1024px) {
      flex-basis: 100%;
    }
  }

  &__input {
    margin: 8px 0;
    font-size: 1.3rem;
    padding: 10px 14px;
    border: 2px solid var(--color-secondary);
    border-radius: 16px;

    @media (min-width: 1024px) {
      flex-basis: 67%;
    }

    &--error {
      border: 2px solid var(--color-tertiary);
    }
  }

  &__validation {
    font-size: 1.2rem;
    margin-bottom: 8px;
    margin-left: 10px;
    color: var(--color-tertiary);

    @media (min-width: 1024px) {
      order: 2;
      font-size: 1.3rem;
      margin: 0 4px;
    }
  }

  &__cta {
    background-color: var(--color-primary);
    color: var(--color-background);
    font-weight: var(--font-weight-medium);
    font-size: 1.3rem;
    height: 38px;
    border-radius: 16px;
    box-shadow: 0px 2px 10px 0px var(--color-primary);

    @media (min-width: 1024px) {
      flex-basis: 30%;
    }
  }
}
