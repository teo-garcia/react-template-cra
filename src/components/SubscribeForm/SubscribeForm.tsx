import './SubscribeForm.scss';
import * as React from 'react';
import { SubscribeFormProps } from '../../tools/types';
import { Button } from '../Button/Button';
import classNames from 'classnames';

function validateEmail(email: string): boolean {
  return /^\S+@\S+$/.test(email);
}

function SubscribeForm(props: SubscribeFormProps) {
  const { label, name, onSubmit } = props;
  const [error, setError] = React.useState(false);
  const [value, setValue] = React.useState('');

  function handleChange(evt: React.ChangeEvent<HTMLInputElement>) {
    const { value } = evt.target;
    setError(false);
    setValue(value);
  }

  function handleSubmit(evt: React.FormEvent<HTMLFormElement>) {
    evt.preventDefault();
    const isValid = validateEmail(value);

    if (isValid) {
      onSubmit(value);
      return;
    }

    setError(true);
  }

  return (
    <form className="subscribe-form" onSubmit={handleSubmit}>
      <label className="subscribe-form__label" htmlFor={name}>
        {label}
      </label>
      <input
        className={classNames('subscribe-form__input', {
          'subscribe-form__input--error': error,
        })}
        name={name}
        id={name}
        onChange={handleChange}
        value={value}
        aria-describedby="error-email"
      />
      {error ? (
        <span id="error-email" className="subscribe-form__validation">
          Please enter a valid email
        </span>
      ) : null}
      <Button className="subscribe-form__cta">Notify Me</Button>
    </form>
  );
}

export { SubscribeForm };
