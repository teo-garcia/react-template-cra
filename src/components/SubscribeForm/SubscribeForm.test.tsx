import { render, fireEvent, screen } from '@testing-library/react';
import { SubscribeForm } from './SubscribeForm';

describe('<SubscribeForm /> tests', function () {
  test('Should match with the component snapshot', function () {
    const { asFragment } = render(
      <SubscribeForm label="Email Address" name="email" onSubmit={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Should call onSubmit prop when is submitted', function () {
    const onMockSubmit = jest.fn();
    render(
      <SubscribeForm
        label="Email Address"
        name="email"
        onSubmit={onMockSubmit}
      />
    );

    fireEvent.change(screen.getByLabelText(/email address/i), {
      target: { value: 'john@doe.com' },
    });
    fireEvent.click(screen.getByText(/notify me/i));
    expect(onMockSubmit).toHaveBeenCalledWith('john@doe.com');
  });

  test('Should show an error when the form is invalid', function () {
    const onMockSubmit = jest.fn();
    render(
      <SubscribeForm
        label="Email Address"
        name="email"
        onSubmit={onMockSubmit}
      />
    );

    fireEvent.change(screen.getByLabelText(/email address/i), {
      target: { value: 'johndoe.com' },
    });
    fireEvent.click(screen.getByText(/notify me/i));
    expect(screen.getByText(/please enter a valid email/i)).toBeInvalid();
    expect(onMockSubmit).not.toHaveBeenCalled();
  });
});
