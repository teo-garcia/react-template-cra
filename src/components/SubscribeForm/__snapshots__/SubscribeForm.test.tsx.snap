// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubscribeForm /> tests Should match with the component snapshot 1`] = `
<DocumentFragment>
  <form
    class="subscribe-form"
  >
    <label
      class="subscribe-form__label"
      for="email"
    >
      Email Address
    </label>
    <input
      aria-describedby="error-email"
      class="subscribe-form__input"
      id="email"
      name="email"
      value=""
    />
    <button
      class="button subscribe-form__cta"
    >
      Notify Me
    </button>
  </form>
</DocumentFragment>
`;
